### PR TITLE
Add atomic update extension functions to MutableStateFlow

### PR DESCRIPTION
**Pull request by lowasser**
Addresses https://github.com/Kotlin/kotlinx.coroutines/issues/2720.

### CODE REVIEW

File: kotlinx-coroutines-core/common/test/flow/sharing/StateFlowTest.kt [hunk 1]
```diff
@@ -193,4 +193,59 @@ class StateFlowTest : TestBase() {
...
+        when (j2.await()) {
+            5 -> assertEquals(3, j3.await())
+            3 -> assertEquals(5, j3.await())
+            else -> fail()
+        }
+        assertEquals(5, state.value)
+    }
+
+    @Test
+    fun update() = runTest {
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/common/test/flow/sharing/StateFlowTest.kt:244]:
```suggestion
    fun testUpdate() = runTest {
```

File: kotlinx-coroutines-core/common/src/flow/StateFlow.kt [hunk 2]
```diff
@@ -186,6 +186,56 @@ public interface MutableStateFlow<T> : StateFlow<T>, MutableSharedFlow<T> {
...
+    }
+}
+
+/**
+ * Updates the [MutableStateFlow.value] atomically using the specified [function] of its value, and returns its
+ * prior value.
+ *
+ * [function] may be evaluated multiple times, if [value] is being concurrently updated.
+ */
+public inline fun <T> MutableStateFlow<T>.updateAndGet(function: (T) -> T): T {
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/common/src/flow/StateFlow.kt:213]: This function seems to be copy-pasted and thus the code does not compile. Have you meant `getAndUpdate` instead?

**Review by qwwdfsad — CHANGES_REQUESTED** [review]
Good job! Please don't forget to run `./gradlew apiDump` as well

**Comment by lowasser** [comment]
Done!

**Comment by qwwdfsad** [comment]
Merged manually as part of #2729, thanks!

I rewrote the tests (they were single-threaded as opposed to their name and also failed on JS) and changed the base branch to `develop`. We have a comprehensive contribution guide here that cover such common pitfalls: https://github.com/Kotlin/kotlinx.coroutines/blob/master/CONTRIBUTING.md#submitting-prs

**Pull request closed without merging**